lib/spack/spack/binary_distribution.py

```diff
@@ -100,438 +80,368 @@ class BinaryCacheIndex(object):
-            fetched_hash = codecs.getreader("utf-8")(fs).read()
-        except (URLError, web_util.SpackWebError) as url_err:
-            errors.append(
-                RuntimeError(
-                    "Unable to read index hash {0} due to {1}: {2}".format(
-                        hash_fetch_url, url_err.__class__.__name__, str(url_err)
-                    )
-                )
-            )
-
-        # The only case where we'll skip attempting to fetch the buildcache
-        # index from the mirror is when we already have a hash for this
-        # mirror, we were able to retrieve one from the mirror, and
-        # the two hashes are the same.
-        if expect_hash and fetched_hash:
-            if fetched_hash == expect_hash:
-                tty.debug("Cached index for {0} already up to date".format(mirror_url))
-                return False
-            else:
-                # We expected a hash, we fetched a hash, and they were not the
-                # same.  If we end up fetching an index successfully and
-                # replacing our entry for this mirror, we should clean up the
-                # existing cache file
-                if mirror_url in self._local_index_cache:
-                    existing_entry = self._local_index_cache[mirror_url]
-                    old_cache_key = existing_entry["index_path"]
-
-        tty.debug("Fetching index from {0}".format(index_fetch_url))
+            with self._index_file_cache.write_transaction(meta_key) as (oldmeta_f, newmeta_f):
+                metadata = {}
+                if oldmeta_f is not None:
+                    try:
+                        metadata = json.load(oldmeta_f)
+                    except ValueError:
+                        # Corrupt metadata, ignore
+                        pass
+
+                # Fast path: someone else updated the index in the moment when we weren't
+                # holding a lock. Our expected_hash may be out-of-date now, so assume
+                # the other guy did good instead of aggressively checking.
+                if mirror_url in metadata:
+                    current_hash = metadata[mirror_url]["hash"]
+                    if current_hash != previous_hash:
+                        return False
+
+                # At this point: we need to fetch the index.json. No two ways about it.
+                # Do the fetch now, so that if it fails we don't leave any cruft behind.
+                # We will catch the exception this raises outside the write_transaction
+                _, _, fs = web_util.read_from_url(index_fetch_url)
+                data = codecs.getreader("utf-8")(fs).read()
+
+                # Decide on the cache key for the data blob, if one hasn't been chosen
+                if mirror_url in metadata:
+                    data_key = metadata[mirror_url]["key"]
+                else:
+                    data_key = newdata_key
+                    if len(metadata) > 0:
```

Comment:
Please just keep it simple. Content hash is enough.

Reply:
Content hashes make this complex, the new implementation uses URL hashes which are far simpler. Quoting part of the updated OP:
> The cache keys used are based on a hash of the mirror's URL, as opposed to alternative approaches:
> - Mirror names are not necessarily consistent across parallel or subsequent Spack invocations, one Spack may refer to the same mirror URL via a different name or vice versa.
> - Content hashes change over time, requiring the removal of out-of-date cache keys during normal operations. Changing keys significantly complicates the parallelism for little real gain.